[serve][llm] add cpu support to ray serve

srinarayan-srikanthan opened the pull request:
## Description
Enable serving of vllm on cpu when the accelerator provided is none, or explicitly there are cpu workers only in a cluster.

## Related issues
"Fixes #53603 ","Related to #56636 ".

## Additional information

enabled mp runtine for cpu and use_gpu by default when accelerator not provided to false.
In addition to the changes need to build vllm for cpu instead of vllm pip package as shown in the vllm installation page : https://docs.vllm.ai/en/latest/getting_started/installation/cpu.html#intelamd-x86

Commit:
add cpu suuport to ray serve

Signed-off-by: root <root@ip-10-0-11-206.ec2.internal>

Commit:
add logger

Signed-off-by: root <root@ip-10-0-11-206.ec2.internal>

Commit:
modified validation logic

Signed-off-by: root <root@ip-10-0-11-206.ec2.internal>

Commit:
fix typo

Signed-off-by: root <root@ip-10-0-11-206.ec2.internal>

eicherseiji commented on python/ray/llm/_internal/serve/engines/vllm/vllm_models.py:
```diff
@@ -265,8 +271,8 @@ def use_gpu(self) -> bool:
         if not self.accelerator_type:
-            # By default, GPU resources are used
-            return True
+            # Use cpu if gpu not provided or none provided
```
I don’t think we want to change the default here. We can either add a “NONE” accelerator type, or add “use_cpu” to the LLMConfig and pipe it through.

I’m leaning toward adding a `NONE` accelerator type and checking for it here.

eicherseiji replied:
`self.accelerator_type` isn't necessarily an indicator of what hardware is on the cluster, it's more of a request for a certain hardware type. TBH I'm not the biggest fan of it, I think it's better replaced by `placement_group_config`.

But in the meantime, if `accelerator_type` is left unset, we want to preserve the behavior that GPU is assumed. CPU support is not included in the basic vLLM install, or even prebuilt wheel, so I think this is reasonable.

So, @srinarayan-srikanthan I think we should add an explicit `use_gpu` field to `LLMConfig` https://github.com/ray-project/ray/blob/fedb6a780a3cc372de89a986e86e3ffdb1e224b1/python/ray/llm/_internal/serve/core/configs/llm_config.py#L141

By default it should be set to None, and can be passed down to `VLLMEngineConfig` to be used to override this propert.

eicherseiji commented:
Thanks for the PR! Would you be open to including some docs (basically just record the steps to test) so folks know how to use the feature?

louie-tsai reviewed (approved):
looks good to me. thanks for the effort to enable CPU for Ray Serve

nrghosh reviewed (commented):
Thanks @srinarayan-srikanthan, +1 to Seiji's last comment

If you run `./ci/lint/lint.sh pre_commit code_format`, we can unblock the unit tests and get them running. (see: https://buildkite.com/ray-project/microcheck/builds/30264/steps/table) 

Thanks for your contribution

kouroshHakha reviewed (commented):
I think the most straightforward and backward compatible way to support CPU is to add a new accelerator_type="CPU" to the list and treat it as special. 

Changing accelerator_type=None from meaning GPU to non GPU is a backward compatibility break that is not desired.

srinarayan-srikanthan committed:
Merge branch 'ray-project:master' into enable-cpu

srinarayan-srikanthan committed:
Add CPU support configuration and validation logic

- Add use_cpu field to LLMConfig for explicit CPU inference control
- Update VLLMEngineConfig validation to handle CPU inference logic
- Allow models to run on CPU when use_cpu=True is specified

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

eicherseiji commented:
@srinarayan-srikanthan try these steps: https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c
<img width="419" height="158" alt="Screenshot 2025-12-11 at 11 24 38 AM" src="https://github.com/user-attachments/assets/73217b43-b482-4a15-9dd2-7098b2be0f94" />

srinarayan-srikanthan committed:
Fix distributed_executor_backend logic for CPU mode

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan committed:
Fix missing use_cpu field in VLLMEngineConfig

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan commented on python/ray/llm/_internal/serve/engines/vllm/vllm_models.py:
```diff
@@ -265,8 +271,8 @@ def use_gpu(self) -> bool:
-            # By default, GPU resources are used
-            return True
+            # Use cpu if gpu not provided or none provided
+            return False
```
done

srinarayan-srikanthan committed:
Fix placement bundles for CPU-only mode

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan committed:
add use_cpu precedence over accelerator_type

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan committed:
Handle explicit use_cpu=False to force GPU mode

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan committed:
Fix code formatting for logger warnings

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan commented:
@eicherseiji  and @kouroshHakha please help review and merge changes

nrghosh reviewed (commented):
defer to @eicherseiji on style, having a top level param makes sense to me vs. overriding/adding to accelerator_type when it's not an accelerator. 

Thanks for the contribution @srinarayan-srikanthan

kouroshHakha reviewed (commented):
@srinarayan-srikanthan thanks for contribution. Some small comments that need addressing before merging:

srinarayan-srikanthan commented on python/ray/llm/_internal/serve/engines/vllm/vllm_models.py:
```diff
@@ -125,21 +131,27 @@ def get_initialization_kwargs(self) -> dict:
+        if not self.use_gpu:
+            # For CPU mode, always use "mp" backend
+            engine_kwargs["distributed_executor_backend"] = "mp"
+            logger.warning("install vllm package for cpu to ensure seamless execution")
```
vllm cpu pip package was not available earlier, but this weekend it was merged, so vllm wheels will support cpu. This is not needed , ray package when installing vllm should handle this. We can create a separate pr on that  if ray does not handle that. Comments? @kouroshHakha

srinarayan-srikanthan replied:
this is the PR that got merged, https://github.com/vllm-project/vllm/pull/28848 , i need to verify but given it got merged it should be.

srinarayan-srikanthan committed:
Improve use_cpu precedence logic with cleaner implementation

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

eicherseiji commented:
Btw @srinarayan-srikanthan have you been able to test this?

srinarayan-srikanthan commented:
@eicherseiji yes, i was able to validate the changes and was able to serve models on cpu

srinarayan-srikanthan committed:
Add basic tests validating use_cpu field functionality and precedence over accelerator_type

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan commented:
@eicherseiji I have added two basic checks, please check if this looks good.

eicherseiji reviewed (approved):
lgtm, thanks. @kouroshHakha in the future, we can consider removing the internal `use_gpu` flag:
https://github.com/ray-project/ray/blob/bd18bf8892d4170a3bc92bdb6f9cc972850aaddc/python/ray/llm/_internal/serve/engines/vllm/vllm_models.py#L264

srinarayan-srikanthan committed:
Lint

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

srinarayan-srikanthan commented:
thank you @eicherseiji , made the changes, should pass the test now.

srinarayan-srikanthan committed:
Fix accelerator hint inconsistency in CPU mode for custom placement groups

Signed-off-by: srinarayan-srikanthan <srinarayan.srikanthan@intel.com>

kouroshHakha reviewed (commented):
could we add some release tests on cpu-only machines @eicherseiji ?

CaoimheCahill commented:
> <img alt="Screenshot 2025-12-16 at 6 24 59 PM" width="1840" height="1191" src="https://private-user-images.githubusercontent.com/58963096/527359911-45a900ae-be50-4bdc-a3aa-be9225dd66b2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzc1NDY2NDIsIm5iZiI6MTc3NzU0NjM0MiwicGF0aCI6Ii81ODk2MzA5Ni81MjczNTk5MTEtNDVhOTAwYWUtYmU1MC00YmRjLWEzYWEtYmU5MjI1ZGQ2NmIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDMwVDEwNTIyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzYTUzMWQ2M2IyNDczMDRkYjg3NDUwNWJjMDU3YWE3ZGEzMjEwNGZmYjk1Y2Y4MTU4MzM4ZjFjZjViNDIxZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.7mO1ZzusevmPg8aR0lneZxlf9fuD08qO57CVaOE_aYY">
> @kouroshHakha FYI the change works well with `vllm==0.12.0+cpu` on an EC2 `m5` instance with the following Ray Serve LLM config:
> 
> ```python
> from ray import serve
> from ray.serve.llm import LLMConfig, build_openai_app
> 
> llm_config = LLMConfig(
>     model_loading_config={
>         "model_id": "qwen-0.5b",
>         "model_source": "Qwen/Qwen2.5-0.5B-Instruct",
>     },
>     engine_kwargs={"dtype": "float16"},  
>     use_cpu=True,
> )
> 
> app = build_openai_app({"llm_configs": [llm_config]})
> serve.run(app, blocking=True)
> ```
> 
> Install command:
> 
> ```
> VLLM_USE_PRECOMPILED=1 VLLM_PRECOMPILED_WHEEL_VARIANT=cpu VLLM_TARGET_DEVICE=cpu uv pip install --editable . --system --extra-index-url https://download.pytorch.org/whl/cpu --index-strategy unsafe-best-match
> ```

Hi, I'm trying to run this on a minikube KubeRay cluster on CPU but this set up doesn't seem to work for me? I keep getting import errors and compatibility issues.
